diem-move/diem-framework/experimental/sources/NFTGallery.move

```diff
@@ -0,0 +1,121 @@
+            // If `owner` already has a token with the same id, remove it from the gallery
+            // and join it with the new token.
+            let original_token = remove_from_gallery<TokenType>(owner, &id);
+            NFT::join<TokenType>(&mut token, original_token);
+        };
+        let gallery = &mut borrow_global_mut<NFTGallery<TokenType>>(owner).gallery;
+        Vector::push_back(gallery, token)
+    }
+
+    /// Remove a token of certain id from the owner's gallery and return it.
+    fun remove_from_gallery<TokenType: store>(owner: address, id: &GUID::ID): Token<TokenType>
+    acquires NFTGallery {
+        assert!(exists<NFTGallery<TokenType>>(owner), EBALANCE_NOT_PUBLISHED);
+        let gallery = &mut borrow_global_mut<NFTGallery<TokenType>>(owner).gallery;
+        let index_opt = index_of_token<TokenType>(gallery, id);
+        assert!(Option::is_some(&index_opt), Errors::limit_exceeded(EID_NOT_FOUND));
+        Vector::remove(gallery, Option::extract(&mut index_opt))
+    }
+
+    /// Finds the index of token with the given id in the gallery.
+    fun index_of_token<TokenType: store>(gallery: &vector<Token<TokenType>>, id: &GUID::ID): Option<u64> {
+        let i = 0;
+        let len = Vector::length(gallery);
+        while (i < len) {
+            if (NFT::id<TokenType>(Vector::borrow(gallery, i)) == *id) {
+                return Option::some(i)
+            };
+            i = i + 1;
+        };
+        Option::none()
+    }
+
+    /// Returns whether the owner has a token with given id.
+    public fun has_token<TokenType: store>(owner: address, token_id: &GUID::ID): bool acquires NFTGallery {
+        Option::is_some(&index_of_token(&borrow_global<NFTGallery<TokenType>>(owner).gallery, token_id))
+    }
+
+    public fun get_token_balance<TokenType: store>(owner: address, token_id: &GUID::ID
+    ): u64 acquires NFTGallery {
+        let gallery = &borrow_global<NFTGallery<TokenType>>(owner).gallery;
+        let index_opt = index_of_token<TokenType>(gallery, token_id);
+
+        if (Option::is_none(&index_opt)) {
+            0
+        } else {
+            let index = Option::extract(&mut index_opt);
+            NFT::balance(Vector::borrow(gallery, index))
+        }
+    }
+
+    /// Transfer `amount` of token with id `GUID::id(creator, creation_num)` from `owner`'s
+    /// balance to `to`'s balance. This operation has to be done by either the owner or an
+    /// approved operator of the owner.
+    public(script) fun transfer_token_between_galleries<TokenType: store>(
+        account: signer,
+        to: address,
+        amount: u64,
+        creator: address,
+        creation_num: u64
+    ) acquires NFTGallery {
+        let owner = Signer::address_of(&account);
+
+        assert!(amount > 0, EINVALID_AMOUNT_OF_TRANSFER);
+
+        // Remove NFT from `owner`'s gallery
+        let id = GUID::create_id(creator, creation_num);
+        let token = remove_from_gallery<TokenType>(owner, &id);
```

Comment:
Same here: I think it's going to be pretty inefficient since we're doing unnecessary `Vector::remove()` and `Vector::push_back()` operations.